### PR TITLE
Functional Tests For HTTP Trace Followed By AspNet* Triggers

### PR DESCRIPTION
**Pull request by kkeirstead**
Added two test cases that exercise the API followed by a collection rule; these tests will fail (and are currently turned off) until https://github.com/dotnet/diagnostics/pull/3425 is integrated.

### CODE REVIEW

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/HttpApi/ApiClientExtensions.cs [hunk 1]
```diff
@@ -287,6 +287,23 @@ public static async Task<ResponseStreamHolder> CaptureLogsAsync(this ApiClient c
             return await client.CaptureLogsAsync(pid, duration, configuration, logFormat, timeoutSource.Token);
         }
 
+        /// <summary>
+        /// GET /trace?pid={pid}&profile={profile}&durationSeconds={duration}
+        /// </summary>
+        public static Task<ResponseStreamHolder> CaptureTraceAsync(this ApiClient client, int pid, TimeSpan duration, TraceProfile? profile)
```
Comment (wiktork) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/HttpApi/ApiClientExtensions.cs:293]: Maybe unify with EgressTraceAsync?

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectionRuleAndApiTests.cs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+                });
+        }
+
+        /// <summary>
+        /// Validates that an AspNetRequestDuration trigger will fire following an HTTP trace.
+        /// </summary>
+        [Theory(Skip = "These tests will fail until #3425 in the diagnostics repo is checked in.")]
+        //[Theory]
+        [InlineData(DiagnosticPortConnectionMode.Listen)]
+        public async Task CollectionRuleAndApi_AspNetRequestDurationTest(DiagnosticPortConnectionMode mode)
```
Comment (jander-msft) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectionRuleAndApiTests.cs:108]: Update this test to use `ScenarioRunner.SingleTarget`

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectionRuleAndApiTests.cs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+        private const string AppUrl = "http://+:10000";
+        private const string AspNetUrlsKey = "ASPNETCORE_Urls";
+
+        /// <summary>
+        /// Validates that an AspNetResponseStatus trigger will fire following an HTTP trace.
+        /// </summary>
+        [Theory(Skip = "These tests will fail until #3425 in the diagnostics repo is checked in.")]
+        //[Theory]
+        [InlineData(DiagnosticPortConnectionMode.Listen)]
+        public async Task CollectionRuleAndApi_AspNetResponseStatusTest(DiagnosticPortConnectionMode mode)
```
Comment (jander-msft) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectionRuleAndApiTests.cs:54]: nit: Remove `async` and return the `Task` from the `ScenarioRunner.SingleTarget` call.

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectionRuleAndApiTests.cs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+        private async Task<ResponseStreamHolder> CaptureTrace(AppRunner runner, ApiClient client)
+        {
+            int processId = await runner.ProcessIdTask;
+
+            int retryCounter = 0;
+
+            // Repeatedly check if dotnet-monitor is detecting our process; due to timing issues,
+            // tests were sometimes failing due to the target process not being found when collecting a trace.
+            while (retryCounter < 5)
```
Comment (jander-msft) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectionRuleAndApiTests.cs:212]: Why is this needed? The `GetProcessesWithRetryAsync` call already takes care of retries.

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectionRuleAndApiTests.cs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests
+{
+    [TargetFrameworkMonikerTrait(TargetFrameworkMonikerExtensions.CurrentTargetFrameworkMoniker)]
+    [Collection(DefaultCollectionFixture.Name)]
+    public class CollectionRuleAndApiTests
```
Comment (jander-msft) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectionRuleAndApiTests.cs:30]: Can we rename this to the class of tests that exist here, maybe `EventSourceInterferenceTests` or `OverlappingEventSourceTests`.

File: src/Tests/Microsoft.Diagnostics.Monitoring.TestCommon/Runners/LoggingRunnerAdapter.cs [hunk 6]
```diff
@@ -17,7 +17,7 @@ public sealed class LoggingRunnerAdapter : IAsyncDisposable
         private readonly ITestOutputHelper _outputHelper;
```
Comment (jander-msft) [inline src/Tests/Microsoft.Diagnostics.Monitoring.TestCommon/Runners/LoggingRunnerAdapter.cs:17]: I think the changes in this file can be undone.

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/OverlappingEventSourceTests.cs [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+            string ExpectedFilePath = Path.Combine(tempDirectory.FullName, "file.txt");
+            string ExpectedFileContent = Guid.NewGuid().ToString("N");
+            string[] urlPaths = new string[] { "", "/Privacy", "" };
+
+            Task ruleStartedTask = null;
+            return ScenarioRunner.SingleTarget(
+                _outputHelper,
+                _httpClientFactory,
+                mode,
+                TestAppScenarios.AspNetSlow.Name,
```
Comment (kkeirstead) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/OverlappingEventSourceTests.cs:65]: nit: this scenario is getting lumped in with AspNetSlow, despite being for ResponseStatus. Should these be split out, or switched to a more appropriate name that applies to all of our AspNet* triggers.

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/OverlappingEventSourceTests.cs [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        {
+            _httpClientFactory = serviceProviderFixture.ServiceProvider.GetService<IHttpClientFactory>();
+            _outputHelper = outputHelper;
+        }
+
+#if NET6_0_OR_GREATER
+        private const string DefaultRuleName = "FunctionalTestRule";
+        private readonly TimeSpan TraceDuration = Timeout.InfiniteTimeSpan;
+        private const string ClientBaseUrl = "http://localhost:10000";
+        private const string AppUrl = "http://+:10000";
```
Comment (kkeirstead) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/OverlappingEventSourceTests.cs:43]: Do we need to change this?

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectionRuleAndApiTests.cs (hunk 3 above)
Comment (jander-msft) [reply src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectionRuleAndApiTests.cs]: We discussed offline... the `Task` needs to be awaited so that the `TemporaryDirectory` instances are not disposed before the body of the test finishes.

File: src/Tests/Microsoft.Diagnostics.Monitoring.TestCommon/TraceTestUtilities.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Microsoft.Diagnostics.Monitoring.TestCommon
+{
+    public static class TraceTestUtilities
+    {
+        public static async Task ValidateTrace(Stream traceStream)
+        {
+            using CancellationTokenSource cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(30));
```
Comment (wiktork) [inline src/Tests/Microsoft.Diagnostics.Monitoring.TestCommon/TraceTestUtilities.cs:18]: We have a central place for all timeouts, maybe add it there?

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/HttpApi/ApiClient.cs [hunk 10]
```diff
@@ -402,6 +402,55 @@ private async Task<ResponseStreamHolder> CaptureLogsAsync(HttpMethod method, str
...
+                CreateTraceUriString(processQuery, duration, profile),
+                content: null,
+                token);
+        }
+
+        private async Task<ResponseStreamHolder> CaptureTraceAsync(HttpMethod method, string uri, HttpContent content, CancellationToken token)
+        {
+            string contentType = ContentTypes.ApplicationOctetStream;
+
+            using HttpRequestMessage request = new(method, uri);
```
Comment (wiktork) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/HttpApi/ApiClient.cs:426]: I'm wondering if long term we could have more shared code between EgressTrace and CaptureTrace but I don't think we need to worry about that now.

**Review by jander-msft — APPROVED** [review]
